### PR TITLE
Rewrite most of the set API to use overloads

### PR DESCRIPTION
**Pull request by guilhermeleobas**
Methods changed:
- set.copy
- set.clear
- set.discard
- set.pop
- set.remove
- set.difference_update
- set.intersection_update
- set.symmetric_difference_update
- set.isdisjoint
- set.issubset
- set.issuperset
- set.union
- set.symetryc_difference
- set.intersection
- set.difference
- operator.sub
- operator.and_
- operator.xor
- operator.or_
- operator.iand
- operator.ior
- operator.isub
- operator.ixor
- operator.le
- operator.gt
- operator.lt
- operator.ge
- operator.ne
- operator.eq

### CODE REVIEW

File: numba/cpython/setobj.py [hunk 1]
```diff
@@ -1351,24 +1406,60 @@ def set_difference_update(context, builder, sig, args):
...
+@intrinsic
+def _set_difference_update(typingctx, a, b):
+    sig = types.none(a, b)
+    return sig, set_difference_update
+
+
+@overload_method(types.Set, "difference_update")
+def set_difference_update_impl(a, b):
+    if not all([isinstance(typ, types.Set) for typ in (a, b)]):
+        return
```
Comment (stuartarchibald) [inline numba/cpython/setobj.py:1419]: Factor this out as function `all_sets(*args)` ?

File: numba/cpython/setobj.py [hunk 2]
```diff
@@ -1298,15 +1298,26 @@ def set_add(context, builder, sig, args):
 
     return context.get_dummy_value()
 
-@lower_builtin("set.discard", types.Set, types.Any)
+
 def set_discard(context, builder, sig, args):
```
Comment (stuartarchibald) [inline numba/cpython/setobj.py:1302]: Move this into the intrinsic `_set_discard` as the `codegen` (and same comment for all the occurrences of this pattern)? i.e.
```python
@intrinsic
def _set_<something>(typingctx, <args>):
    sig = <>
    def codegen(cgctx, builder, sig, args): # This is the `set_<something>` function at present.
        ...
    return sig, codegen
```

File: numba/cpython/setobj.py [hunk 3]
```diff
@@ -1328,21 +1350,54 @@ def set_remove(context, builder, sig, args):
...
 
+@intrinsic
+def _set_remove(typingctx, s, item):
+    sig = types.none(s, item)
+    return sig, set_remove
+
+
+@overload_method(types.Set, "remove")
+def ol_set_remove(s, item):
+    # XXX: should item have the same type as s.dtype?
```
Comment (stuartarchibald) [inline numba/cpython/setobj.py:1361]: I think yes given the current implementation. e.g.:
```python
from numba import njit
import numpy as np

@njit
def foo():
    x = {1, 2, 3, 4}
    x.remove('a')

foo()
```
ends up with:
```
TypeError: Type of #3 arg mismatch: i64 != {i8*, i64, i32, i32, i64, i8*, i8*}
```
As the reflected set implementation is pending deprecation I don't think there's any need to add support for this, but the issue in lowering can be prevented at type inference time so it should be. Suggest fixing the typing and adding a test.

File: numba/core/typing/setdecl.py [hunk 4]
```diff
@@ -97,18 +67,6 @@ def _resolve_xxx_update(self, set, args, kws):
         if isinstance(iterable, types.Set) and iterable.dtype == set.dtype:
             return signature(types.none, iterable)
 
-    @bound_function("set.difference_update")
-    def resolve_difference_update(self, set, args, kws):
-        return self._resolve_xxx_update(set, args, kws)
```
Comment (stuartarchibald) [inline numba/core/typing/setdecl.py:102]: Is `self._resolve_xxx_update` now dead code?

File: numba/cpython/setobj.py (hunk 1 above)
Comment (stuartarchibald) [reply numba/cpython/setobj.py:1419]: Also, the original `_resolve_xxx_update` method requires that the `dtype` of the two sets is the same:
https://github.com/numba/numba/blob/4bc04ceb28a3bce0461e23db16b8826745e96aef/numba/core/typing/setdecl.py#L92-L98

File: numba/cpython/setobj.py [hunk 5]
```diff
@@ -1399,34 +1490,53 @@ def set_update(context, builder, sig, args):
...
 
 # Set operations creating a new set
 
-@lower_builtin(operator.sub, types.Set, types.Set)
-@lower_builtin("set.difference", types.Set, types.Set)
-def set_difference(context, builder, sig, args):
+@overload(operator.sub)
+@overload_method(types.Set, "difference")
+def impl_set_difference(a, b):
+    if not all([isinstance(typ, types.Set) for typ in (a, b)]):
```
Comment (stuartarchibald) [inline numba/cpython/setobj.py:1524]: Think these also need `dtype` checks. Suggest checking all of these to make sure the `dtype` check is present.

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch. This refactoring seems broadly correct on inspection. There's a few minor things to resolve particularly in relation to ensuring `dtype` checks are made, once resolved should be good to merge.

Note: There should be no impact on the CUDA target as `set` isn't supported due to the use of the NRT.

Thanks again!

File: numba/core/typing/setdecl.py (hunk 4 above)
Comment (guilhermeleobas) [reply numba/core/typing/setdecl.py:102]: Yes, it is! I'll remove it

File: numba/cpython/setobj.py (hunk 2 above)
Comment (guilhermeleobas) [reply numba/cpython/setobj.py:1302]: Ok! I kept this way to reduce the patch size and to make it easier to review.

File: numba/cpython/setobj.py (hunk 3 above)
Comment (stuartarchibald) [reply numba/cpython/setobj.py]: Thanks for making the change, I think it would be good to add a test for it.

Comment (guilhermeleobas) [reply numba/cpython/setobj.py]: Oops, sorry. I forgot to include a test for it

File: numba/cpython/setobj.py [hunk 6]
```diff
@@ -103,6 +103,11 @@ def is_hash_used(context, builder, h):
     return builder.icmp_unsigned('<', h, deleted)
 
 
+def all_set(*args):
+    return all([isinstance(typ, types.Set) for typ in args]) and \
+        all([args[0].dtype == s.dtype for s in args])
```
Comment (stuartarchibald) [inline numba/cpython/setobj.py:108]: I just ran the unit tests for this and found that whilst a mixed dtype operation is correctly rejected, the error message does not describe the problem very well, it just reports `No implementation of function...<some function>... found for signature...` with no indication as to the actual problem. Maybe this function should:
* raise something like `TypingError(f"All arguments must be Sets, got {args}")` if the types are not all `Set`s.
* raise  something like `TypingError(f"All Sets must be of the same type, got {args}")` if the dtypes are not all the same.

What do you think?

Comment (stuartarchibald) [reply numba/cpython/setobj.py]: Thanks for adding the differentiating error messages in 5510cb6, any chance you could also please add test(s) to demonstrate that each path is correctly taken, i.e. check the correct exceptions are raised for the known erroneous input? Many thanks.

Comment (guilhermeleobas) [reply numba/cpython/setobj.py]: Done!

File: numba/tests/test_sets.py [hunk 7]
```diff
@@ -171,36 +171,25 @@ def union_usecase(a, b):
...
-    """ % dict(op=op)
-    return compile_function('inplace_operator_usecase', code, globals())
-
-def make_comparison_usecase(op):
-    code = """if 1:
-    def comparison_usecase(a, b):
-        return set(a) %(op)s set(b)
-    """ % dict(op=op)
-    return compile_function('comparison_usecase', code, globals())
+def union_usecase__2(a, b):
```
Comment (stuartarchibald) [inline numba/tests/test_sets.py:186]: These `__2` function names don't really note their intended use. Given they are used only in the unit tests in `TestSetsInvalid` perhaps just define them locally in those unit tests?

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the updates @guilhermeleobas, couple of minor things to look at else looks good.

**Review by stuartarchibald — APPROVED** [review]
Thanks for the patch and fixes.

**Pull request merged**